File: src/main/java/com/lumiyaviewer/lumiya/cloud/common/LogMessageBatch.java
package com.lumiyaviewer.lumiya.cloud.common;

import android.os.Bundle;
import android.os.Parcelable;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableList.Builder;
import java.util.Collection;
import java.util.List;
import java.util.UUID;

public class LogMessageBatch implements Bundleable {
    public final String agentName;
    public final UUID agentUUID;
    public final long lastMessageID;
    public final ImmutableList<LogChatMessage> messages;

    public LogMessageBatch(Bundle bundle) {
        this.agentUUID = UUID.fromString(bundle.getString("agentUUID"));
        this.agentName = bundle.getString("agentName");
        this.lastMessageID = bundle.getLong("lastMessageID");
        Parcelable[] parcelableArray = bundle.getParcelableArray("messages");
        Builder builder = ImmutableList.builder();
        if (parcelableArray != null) {
            for (Parcelable parcelable : parcelableArray) {
                if (parcelable instanceof Bundle) {
                    builder.add(new LogChatMessage((Bundle) parcelable));
                }
            }
        }
        this.messages = builder.build();
    }

    public LogMessageBatch(UUID uuid, String str, List<LogChatMessage> list, long j) {
        this.agentUUID = uuid;
        this.agentName = str;
        this.messages = ImmutableList.copyOf((Collection) list);
        this.lastMessageID = j;
    }

    public Bundle toBundle() {
        Bundle bundle = new Bundle();
        bundle.putString("agentUUID", this.agentUUID.toString());
        bundle.putString("agentName", this.agentName);
        bundle.putLong("lastMessageID", this.lastMessageID);
        Parcelable[] parcelableArr = new Bundle[this.messages.size()];
        int i = 0;
        while (true) {
            int i2 = i;
            if (i2 >= this.messages.size()) {
                bundle.putParcelableArray("messages", parcelableArr);
                return bundle;
            }
            parcelableArr[i2] = ((LogChatMessage) this.messages.get(i2)).toBundle();
            i = i2 + 1;
        }
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/cloud/common/LogMessageBatch.kt
package com.lumiyaviewer.lumiya.cloud.common

import android.os.Bundle
import com.google.common.collect.ImmutableList
import java.util.UUID

/**
 * Modern Kotlin LogMessageBatch data class
 * Represents a batch of chat messages for cloud logging
 */
data class LogMessageBatch(
    val agentUUID: UUID,
    val agentName: String,
    val messages: ImmutableList<LogChatMessage>,
    val lastMessageID: Long
) : Bundleable {
    
    /**
     * Creates a LogMessageBatch from a Bundle
     */
    constructor(bundle: Bundle) : this(
        agentUUID = UUID.fromString(bundle.getString("agentUUID") ?: ""),
        agentName = bundle.getString("agentName") ?: "",
        messages = run {
            val builder = ImmutableList.builder<LogChatMessage>()
            bundle.getParcelableArray("messages")?.forEach { parcelable ->
                if (parcelable is Bundle) {
                    builder.add(LogChatMessage(parcelable))
                }
            }
            builder.build()
        },
        lastMessageID = bundle.getLong("lastMessageID")
    )
    
    /**
     * Convenience constructor that accepts a List
     */
    constructor(
        agentUUID: UUID,
        agentName: String,
        messages: List<LogChatMessage>,
        lastMessageID: Long
    ) : this(
        agentUUID = agentUUID,
        agentName = agentName,
        messages = ImmutableList.copyOf(messages),
        lastMessageID = lastMessageID
    )
    
    /**
     * Converts this batch to a Bundle
     */
    override fun toBundle(): Bundle = Bundle().apply {
        putString("agentUUID", agentUUID.toString())
        putString("agentName", agentName)
        putLong("lastMessageID", lastMessageID)
        
        val messageBundles = Array(messages.size) { index ->
            messages[index].toBundle()
        }
        putParcelableArray("messages", messageBundles)
    }
}